File: Design/Stripe/FraudDetection/FraudDetection.java
package FraudDetection;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

public class FraudDetection {
  public List<String> detectFraudulentMerchants(
      String nonFraudCodes,
      String fraudCodes,
      List<String> mccThresholds,
      List<String> accountIdMcc,
      String min,
      List<String> charges) {

    Set<String> fraudCodeSet = new HashSet<>(Arrays.asList(fraudCodes.split(",")));

    Map<String, Double> mccToThresholds = new HashMap<>();
    for (String mccToThreshold : mccThresholds) {
      String mcc = mccToThreshold.split(",")[0];
      double threshold = Double.parseDouble(mccToThreshold.split(",")[1]);
      mccToThresholds.put(mcc, threshold);
    }

    Map<String, String> accountIdToMcc = new HashMap<>();
    for (String accountId : accountIdMcc) {
      String[] parts = accountId.split(",");
      String account = parts[0];
      String mcc = parts[1];
      accountIdToMcc.put(account, mcc);
    }

    Map<String, Integer> accountToTotalCharges = new HashMap<>();
    Map<String, Integer> accountToFraud = new HashMap<>();
    for (String transaction : charges) {
      if (!transaction.startsWith("CHARGE")) continue;

      String[] parts = transaction.split(",");
      String acc = parts[2];
      String code = parts[4];

      accountToTotalCharges.put(acc, accountToTotalCharges.getOrDefault(acc, 0) + 1);
      if (fraudCodeSet.contains(code)) {
        accountToFraud.put(acc, accountToFraud.getOrDefault(acc, 0) + 1);
      }
    }

    // Parse 'min' as double to handle both integer and floating-point values
    double minDouble = Double.parseDouble(min);

    List<String> ret = new ArrayList<>();
    for (String acc : accountToTotalCharges.keySet()) {
      int total = accountToTotalCharges.get(acc);
      int fraud = accountToFraud.getOrDefault(acc, 0);
      String mcc = accountIdToMcc.get(acc);
      double threshold = mccToThresholds.getOrDefault(mcc, 0.0);

      // Compare 'total' with 'minDouble' and fraud count with the threshold
      if (total >= minDouble && fraud >= threshold) {
        ret.add(acc);
      }
    }

    Collections.sort(ret);
    return ret;
  }

  public static void main(String[] args) {
    String nonFraudCodes = "approved,invalid_pin,expired_card"; // Single string with comma-separated values
    String fraudCodes = "do_not_honor,stolen_card,lost_card";   // Single string with comma-separated values
    List<String> mccThresholds = Arrays.asList(
        "retail,0.5", "airline,0.25", "venue,0.25", "restaurant,0.8" // List of strings for mccThresholds
    );
    List<String> merchantMCCs = Arrays.asList(
        "acct_1,airline", "acct_2,venue", "acct_3,venue" // List of strings for merchantMCCs
    );
    String minTransactions = "0"; // The minTransactions as a string that can represent integer or double
    List<String> transactions = Arrays.asList( // List of strings for transactions
        "CHARGE,ch_1,acct_1,100,do_not_honor",
        "CHARGE,ch_2,acct_1,200,approved",
        "CHARGE,ch_3,acct_1,300,do_not_honor",
        "CHARGE,ch_4,acct_2,100,approved",
        "CHARGE,ch_5,acct_2,200,approved",
        "CHARGE,ch_6,acct_1,300,lost_card",
        "CHARGE,ch_7,acct_2,100,approved",
        "CHARGE,ch_8,acct_2,200,approved",
        "CHARGE,ch_9,acct_3,100,approved",
        "CHARGE,ch_10,acct_3,100,approved",
        "CHARGE,ch_11,acct_3,100,approved",
        "CHARGE,ch_12,acct_3,100,approved",
        "CHARGE,ch_13,acct_3,100,stolen_card",
        "CHARGE,ch_14,acct_2,100,stolen_card"
    );

    FraudDetection f = new FraudDetection();
    List<String> fraudulentMerchants = f.detectFraudulentMerchants(
        nonFraudCodes, fraudCodes, mccThresholds, merchantMCCs, minTransactions, transactions
    );
    System.out.println(String.join(",", fraudulentMerchants)); // Should return acct_1, acct_3 based on your criteria
  }
}

File: Design/Stripe/fraud_detection.py
import sys

risk_list = []
safe_list = []
mcc_threshold_map = {}
account_mcc_map = {}
min_number_of_transactions = 0
merchant_map = {}
risk_merchants = set()
transaction_account_map = {}


class Merchant:
    def __init__(self, account_id, mcc, threshold):
        self.account_id = account_id
        self.mcc = mcc
        self.threshold = threshold
        self.total_transactions = 0
        self.risk_transactions = 0

    def risk_transaction(self):
        self.total_transactions += 1
        self.risk_transactions += 1
        self.calculate_risk()

    def safe_transaction(self):
        self.total_transactions += 1
        self.calculate_risk()

    def calculate_risk(self):
        if self.total_transactions >= min_number_of_transactions and self.risk_transactions != 0 and self.total_transactions != 0:
            risk = self.risk_transactions / self.total_transactions
            if risk >= self.threshold:
                if self.account_id not in risk_merchants:
                    risk_merchants.add(self.account_id)
            else:
                if self.account_id in risk_merchants:
                    risk_merchants.remove(self.account_id)

    def dispute_transactions(self):
        self.risk_transactions -= 1
        self.calculate_risk()


def main():
    global min_number_of_transactions  # 声明使用全局变量

    safe_list.extend(sys.stdin.readline().strip().replace('"', '').split(','))
    risk_list.extend(sys.stdin.readline().strip().replace('"', '').split(','))

    for line in sys.stdin:
        parts = line.strip().split(',')
        if parts[0] == 'end':
            print(sorted(risk_merchants))
            return
        if len(parts) == 2:
            first, second = parts[0], parts[1]
            if first[:4] == 'acct':
                account_mcc_map[first] = second
            elif first[:7] == 'DISPUTE':
                account_id = transaction_account_map[second]
                merchant_map[account_id].dispute_transactions()
            else:
                mcc_threshold_map[first] = float(second)
        elif len(parts) == 1:
            first = parts[0]
            min_number_of_transactions = int(first)
        else:
            charge, transaction_id, account_id, amount, transaction_type = parts[0], parts[1], parts[2], parts[3], parts[4]
            transaction_account_map[transaction_id] = account_id
            if account_id not in merchant_map:
                mcc = account_mcc_map[account_id]
                threshold = mcc_threshold_map[mcc]
                new_merchant = Merchant(account_id, mcc, threshold)
                merchant_map[account_id] = new_merchant
            if transaction_type in risk_list:
                merchant_map[account_id].risk_transaction()
            if transaction_type in safe_list:
                merchant_map[account_id].safe_transaction()


if __name__ == "__main__":
    main()

# "approved","invalid_pin","expired_card"
# "do_not_honor","stolen_card"," lost_card"
# retail,0.5
# airline,0.25
# restaurant,0.8
# venue,0.25
# acct_1,airline
# acct_2,venue
# acct_3,venue
# 3
# CHARGE,ch_1,acct_1,100,do_not_honor
# CHARGE,ch_2,acct_1,200,approved
# CHARGE,ch_3,acct_1,300,do_not_honor
# CHARGE,ch_4,acct_2,400,approved
# CHARGE,ch_5,acct_2,500,approved
# CHARGE,ch_6,acct_1,600,lost_card
# CHARGE,ch_7,acct_2,700,approved
# CHARGE,ch_8,acct_2,800,approved
# CHARGE,ch_9,acct_3,800,approved
# CHARGE,ch_10,acct_3,700,approved
# CHARGE,ch_11,acct_3,600,approved
# CHARGE,ch_12,acct_3,500,stolen_card
# CHARGE,ch_13,acct_3,500,stolen_card
# CHARGE,ch_14,acct_2,400,stolen_card


# "approved","invalid_pin","expired_card"
# "do_not_honor","stolen_card","lost_card"
# retail,0.8
# venue,0.25
# acct_1,retail
# acct_2,retail
# 2
# CHARGE,ch_1,acct_1,100,do_not_honor
# CHARGE,ch_2,acct_1,200,lost_card
# CHARGE,ch_3,acct_1,300,do_not_honor
# DISPUTE,ch_2
# CHARGE,ch_4,acct_2,400,lost_card
# CHARGE,ch_5,acct_2,500,lost_card
# CHARGE,ch_6,acct_1,600,lost_card
# CHARGE,ch_7,acct_2,700,lost_card
# CHARGE,ch_8,acct_2,800,do_not_honor